File: Algorithm/Python/981. Time Based Key-Value Store.py
# -*- coding: utf-8 -*-
# Created on Thu Mar 07 2019 20:36:25
# Author: WuLC
# EMail: liangchaowu5@gmail.com

# binary search

from collections import defaultdict
from bisect import bisect_left

class TimeMap(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.db = defaultdict(list)

    def set(self, key, value, timestamp):
        """
        :type key: str
        :type value: str
        :type timestamp: int
        :rtype: None
        """
        self.db[key].append((timestamp, value))

    def get(self, key, timestamp):
        """
        :type key: str
        :type timestamp: int
        :rtype: str
        """
        if key not in self.db:
            return ""
        # custom binary search
        # idx = self.binary_search(key, timestamp)
        # built-in binary search
        idx = bisect_left(self.db[key], (timestamp, '{')) # use '{' because its ascii code is larger than lowercase letter
        return self.db[key][idx-1][1] if idx != 0 else ""

    def binary_search(self, key, timestamp):
        left, right = 0, len(self.db[key]) - 1
        while left < right:
            mid = left + ((right-left)>>1)
            if self.db[key][mid][0] == timestamp:
                break
            elif self.db[key][mid][0] < timestamp:
                left = mid + 1
            else:
                right = mid - 1
        return left if self.db[key][left][0] <= timestamp else left-1



# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)

File: Algorithm/Go/981. Time Based Key-Value Store.go
/*
 * Created on Thu Mar 07 2019 22:21:15
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

// use two hashmap, one for timestamp, the other for value
type TimeMap struct {
	TimeDB  map[string][]int
	ValueDB map[string][]string
}

/** Initialize your data structure here. */
func Constructor() TimeMap {
	return TimeMap{make(map[string][]int), make(map[string][]string)}
}

func (this *TimeMap) Set(key string, value string, timestamp int) {
	if _, present := this.TimeDB[key]; !present {
		this.TimeDB[key] = []int{}
		this.ValueDB[key] = []string{}
	}
	this.TimeDB[key] = append(this.TimeDB[key], timestamp)
	this.ValueDB[key] = append(this.ValueDB[key], value)
}

func (this *TimeMap) Get(key string, timestamp int) string {
	if val, present := this.TimeDB[key]; !present || val[0] > timestamp {
		return ""
	}
	idx := this.BinarySearch(key, timestamp)
	return this.ValueDB[key][idx]
}

func (this TimeMap) BinarySearch(key string, timestamp int) int {
	left, right := 0, len(this.TimeDB[key])-1
	for left < right {
		mid := left + ((right - left) >> 1)
		if this.TimeDB[key][mid] == timestamp {
			left = mid
			break
		} else if this.TimeDB[key][mid] < timestamp {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	if this.TimeDB[key][left] <= timestamp {
		return left
	} else {
		return left - 1
	}
}

/**
 * Your TimeMap object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Set(key,value,timestamp);
 * param_2 := obj.Get(key,timestamp);
 */